### PR TITLE
Pinned Posts

### PR DESCRIPTION
**Pull request by abeddow91**
## What does this change?

This PR adds pinned posts to both liveblogs and deadblogs and is a collaboration between @abeddow91, @joecowton1 and @alinaboghiu.

This PR adds the logic required to render pinned posts on blogs and the UI to render them.

In order to do this we have completed the following tasks:

- [x] added a pinned field to CanonicalLiveBlog which uses the body:pinned CAPI parameter
- [x] filtered the pinned post from the regular event blocks 
- [x] added pinned to the BlockAttributes object
- [x] only filter pinned posts from other blocks if they're the most recent top block
- [x] move the liveblock into its own component so it can be shared by regular and pinned blocks.
- [x] create a pure css collapsible container for the pinned block. The expand/collapse button has been built as toggle rather than a button. This is because it allows us to 'listen' and react to state change in CSS and create a collapsible content container that is fully functional with JavaScript disabled. 
- [x] add event listener to scroll user to the top of the pinned post on collapse - NB this feature only works with JavaScript enabled.
- [x] use dynamic styling to reflect the pillar colour of the blog

NB - There is ongoing Editorial Tools work alongside this which aims to move the pinned post attribute from block to blog. Once this work has been completed, it will need to be reflected here.

## Does this change need to be reproduced in dotcom-rendering ?

- [ ] No
- [x] Yes (please indicate your plans for DCR Implementation)

### Does this affect other platforms?

- [ ] AMP 
- [ ] Apps
- [ ] Other (please specify)

### Does this affect GLabs Paid Content Pages? Should it have support for Paid Content?

- [x] No
- [ ] Yes (please give details)

### Does this change break ad-free?

- [x] No
- [ ] It did, but tests caught it and I fixed it
- [ ] It did, but there was no test coverage so I added that then fixed it

### Does this change update the version of CAPI we're using?

- [] No, all the existing database files are just fine
- [ ] Yes, and I have [re-run all the tests locally and checked in all the updated data/database/xyz files](https://github.com/guardian/frontend/blob/main/docs/03-dev-howtos/15-updating-test-database.md)

### Accessibility test checklist

- [ ] [Tested with screen reader](https://accessibility.gutools.co.uk/testing/web/screen-readers/)
- [ ] [Navigable with keyboard](https://accessibility.gutools.co.uk/testing/web/keyboard-navigation/)
- [x] [Colour contrast passed](https://accessibility.gutools.co.uk/testing/web/colour-contrast/)

### Tested

- [x] Locally
- [ ] On CODE (optional)

### CODE REVIEW

File: static/src/stylesheets/module/content-garnett/_live-blog.scss [hunk 1]
```diff
@@ -369,3 +371,115 @@ input:checked + .live-blog__filter-switch-slider:before {
...
+    @include mq(tablet) {
+        margin-right: $gs-gutter;
+        margin-left: $block-padding-left;
+    }
+
+}
+
+.pinned-block__overlay {
+    background-image: linear-gradient(0deg, #ffffff, #ffffff 40%, transparent);
+    height: 80px;
```
Comment (joecowton1) [inline static/src/stylesheets/module/content-garnett/_live-blog.scss:478]: Should this be rem?

File: dev-build/project/build.properties [hunk 2]
```diff
@@ -0,0 +1 @@
+sbt.version=1.5.5
```
Comment (joecowton1) [inline dev-build/project/build.properties:1]: Do we mean for this to be in the PR?

File: static/src/stylesheets/module/content-garnett/_live-blog.scss (hunk 1 above)
Comment (abeddow91) [reply static/src/stylesheets/module/content-garnett/_live-blog.scss]: good spot, converted to rem!

File: dev-build/project/build.properties (hunk 2 above)
Comment (abeddow91) [reply dev-build/project/build.properties:1]: No that shouldn't be there. I'll remove

File: article/app/controllers/LiveBlogController.scala [hunk 3]
```diff
@@ -5,7 +5,7 @@ import com.gu.contentapi.client.utils.format.{NewsPillar, SportPillar}
 import common.`package`.{convertApiExceptions => _, renderFormat => _}
 import common.{JsonComponent, RichRequestHeader, _}
 import contentapi.ContentApiClient
-import experiments.{ActiveExperiments, Experiment, LiveblogRendering}
+import experiments.{ActiveExperiments, LiveblogPinnedPost, LiveblogRendering}
```
Comment (alinaboghiu) [inline article/app/controllers/LiveBlogController.scala:8]:
```suggestion
import experiments.{ActiveExperiments, LiveblogPinnedBlock, LiveblogRendering}
```

File: article/app/controllers/LiveBlogController.scala [hunk 4]
```diff
@@ -283,12 +281,18 @@ class LiveBlogController(
...
         Left(MinutePage(liveBlog, StoryPackages(liveBlog.metadata.id, response)), blocks)
       case liveBlog: Article if liveBlog.isLiveBlog =>
-        createLiveBlogModel(liveBlog, response, range, filterKeyEvents).left
+        val pinnedPostSwitch = ActiveExperiments.isParticipating(LiveblogPinnedPost)
+        createLiveBlogModel(
+          liveBlog,
+          response,
+          range,
+          filterKeyEvents,
+          pinnedPostSwitch,
```
Comment (alinaboghiu) [inline article/app/controllers/LiveBlogController.scala:290]: pinnedBlockSwitch

File: article/app/model/LiveBlogHelpers.scala [hunk 5]
```diff
@@ -62,6 +62,7 @@ object LiveBlogHelpers {
       response: ItemResponse,
       range: BlockRange,
       filterKeyEvents: Boolean,
+      pinnedPostSwitch: Boolean,
```
Comment (alinaboghiu) [inline article/app/model/LiveBlogHelpers.scala:65]: pinnedBlockSwitch

File: article/app/model/LiveBlogHelpers.scala [hunk 6]
```diff
@@ -101,6 +102,7 @@ object LiveBlogHelpers {
             currentPage = pageModel,
             related = StoryPackages(liveBlog.metadata.id, response),
             filterKeyEvents = filterKeyEvents,
+            pinnedPostSwitch = pinnedPostSwitch,
```
Comment (alinaboghiu) [inline article/app/model/LiveBlogHelpers.scala:105]: pinnedBlockSwitch

File: article/app/views/liveblog/dateBlock.scala.html [hunk 7]
```diff
@@ -1,9 +1,12 @@
 @import model.liveblog.LiveBlogDate
 
-@(date: Option[LiveBlogDate])
-
-@date.map { date =>
-    <time datetime="@date.fullDate" data-relativeformat="med" class=" js-timestamp" itemprop="datePublished">@date.ampm <span class="timezone">@date.gmt</span></time>
-    <span class="block-time__absolute">@date.hhmm</span>
-}
+@(date: Option[LiveBlogDate], showFullDate: Boolean = true, showTimestamp: Boolean = true)
```
Comment (alinaboghiu) [inline article/app/views/liveblog/dateBlock.scala.html:3]: see if it looks better as two separate methods?

File: article/app/views/liveblog/dateBlock.scala.html [hunk 8]
```diff
@@ -1,9 +1,12 @@
...
+    @date.map { date =>
+        @if(showFullDate) {
```
Comment (alinaboghiu) [inline article/app/views/liveblog/dateBlock.scala.html:5]:
```suggestion
        @if(showDate) {
```

File: article/app/views/liveblog/liveBlogBlock.scala.html [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+@import views.BodyProcessor
+@import model.DotcomContentType
+@(block: BodyBlock, article: Article, timezone: DateTimeZone, isPinnedPost: Boolean = false)(implicit request: RequestHeader, context: model.ApplicationContext)
+@*
+* This template is for individual blocks
+*@
+<div
+id="block-@block.id"
+data-sort-time="@block.publishedCreatedTimestamp()"
+class=" @if(!isPinnedPost) {block block--content@block.eventClass@block.attributes.membershipPlaceholder.map { _ => js-insert-epic-after }}"
```
Comment (alinaboghiu) [inline article/app/views/liveblog/liveBlogBlock.scala.html:13]: if(isRegularBlock)

File: article/app/views/liveblog/liveBlogBlocks.scala.html [hunk 10]
```diff
@@ -1,67 +1,9 @@
...
-        }
-        @views.html.fragments.share.blockLevelSharing(
-            s"block-${block.id}",
-            article.sharelinks.elementShares(s"block-${block.id}", None),
-            article.metadata.contentType.getOrElse(DotcomContentType.Unknown),
-            false
-        )
-
-    </div>
+    @views.html.liveblog.liveBlogBlock(block, article, timezone)
```
Comment (alinaboghiu) [inline article/app/views/liveblog/liveBlogBlocks.scala.html:8]:
```suggestion
    @views.html.liveblog.liveblogBlock(block, article, timezone)
```

File: article/app/views/liveblog/liveBlogBody.scala.html [hunk 11]
```diff
@@ -106,6 +106,7 @@ <h1 itemprop="headline" class="content__headline js-score">@Html(article.trail.h
 
                     @fragments.articleAsideSlot(shouldShowAds(model), articleAsideOptionalSizes(false))
 
+
```
Comment (alinaboghiu) [inline article/app/views/liveblog/liveBlogBody.scala.html:109]: undo

File: article/app/views/liveblog/liveBlogBodyContent.scala.html [hunk 12]
```diff
@@ -36,11 +36,12 @@
                 </span>
             </div>
         } else {
+            @if(model.pinnedPostSwitch && model.currentPage.pinnedPost){
```
Comment (alinaboghiu) [inline article/app/views/liveblog/liveBlogBodyContent.scala.html:39]: rename to pinned block

File: article/app/views/liveblog/pinnedPost.scala.html [hunk 13]
```diff
@@ -0,0 +1,28 @@
+@import org.joda.time.DateTimeZone
```
Comment (alinaboghiu) [inline article/app/views/liveblog/pinnedPost.scala.html:1]: rename file and references to pinned block

File: article/app/views/liveblog/pinnedPost.scala.html [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+@import model.Article
+@import model.liveblog.{BodyBlock, LiveBlogDate}
+
+@(block: BodyBlock, article: Article, timezone: DateTimeZone)(implicit request: RequestHeader, context: model.ApplicationContext)
+@*
+* This template is to wrap pinned posts
+*@
+<div class="pinned-block__collapsible-container">
```
Comment (alinaboghiu) [inline article/app/views/liveblog/pinnedPost.scala.html:9]:
```suggestion
<div class="pinned-block--collapsible-container">
```

Comment (alinaboghiu) [inline article/app/views/liveblog/pinnedPost.scala.html:9]:
```suggestion
<div class="pinned-block">
```

File: article/app/views/liveblog/pinnedPost.scala.html [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+    <input id="collapsible" class="pinned-block__toggle" type="checkbox" tabindex="-1" aria-controls="pinned-block" />
```
Comment (alinaboghiu) [inline article/app/views/liveblog/pinnedPost.scala.html:10]:
```suggestion
    <input id="collapsible-button" class="pinned-block__toggle" type="checkbox" tabindex="-1" aria-controls="pinned-block" />
```

File: article/app/views/liveblog/pinnedPost.scala.html [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+    <label for="collapsible" class="pinned-block__btn pinned-block--toggle" tabindex="0">
```
Comment (alinaboghiu) [inline article/app/views/liveblog/pinnedPost.scala.html:11]: try without tab index

File: article/app/views/liveblog/pinnedPost.scala.html (hunk 15 above)
Comment (alinaboghiu) [inline article/app/views/liveblog/pinnedPost.scala.html:10]:
```suggestion
    <input id="collapsible" class="pinned-block__input" type="checkbox" tabindex="-1" aria-controls="pinned-block" />
```

File: article/app/views/liveblog/pinnedPost.scala.html (hunk 16 above)
Comment (alinaboghiu) [reply article/app/views/liveblog/pinnedPost.scala.html:11]: pinned-block--toggle is a bit confusing here, what does it drive?

File: article/app/views/liveblog/pinnedPost.scala.html [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+        @fragments.inlineSvg("plus", "icon", List("pinned-block__btn-icon--expand"))
+        @fragments.inlineSvg("minus", "icon", List("pinned-block__btn-icon--collapse"))
+    </label>
+
+    <section class="pinned-block__collapsible-content">
+        <div class="pinned-block__header">
```
Comment (alinaboghiu) [inline article/app/views/liveblog/pinnedPost.scala.html:17]:
```suggestion
        <div class="pinned-block-content__header">
```

File: static/src/javascripts/bootstraps/enhanced/liveblog.js [hunk 18]
```diff
@@ -59,21 +87,27 @@ const setupListeners = () => {
         const param = `?filterKeyEvents=${hasParam ? 'false' : 'true'}#liveblog-content`;
         window.location.assign(`${window.location.pathname}${param}`);
     })
+
+    bean.on(document.body, 'click', '.pinned-block__btn', () => {
+        const pinnedBlockTop = document.querySelector('.pinned-block__header')
+        const pinnedBlockToggle = document.querySelector('.pinned-block__toggle')
+        pinnedBlockToggle.checked && scrollToElement(pinnedBlockTop)
```
Comment (alinaboghiu) [inline static/src/javascripts/bootstraps/enhanced/liveblog.js:94]: only scroll if not fully visible changes not here

File: article/app/controllers/LiveBlogController.scala (hunk 3 above)
Comment (abeddow91) [reply article/app/controllers/LiveBlogController.scala:8]: adressed in ab/pinned-block-rename

File: article/app/controllers/LiveBlogController.scala (hunk 4 above)
Comment (abeddow91) [reply article/app/controllers/LiveBlogController.scala:290]: addresed in ab/pinned-block-rename

File: article/app/model/LiveBlogHelpers.scala (hunk 5 above)
Comment (abeddow91) [reply article/app/model/LiveBlogHelpers.scala:65]: addressed in ab/pinned-block-rename

File: article/app/model/LiveBlogHelpers.scala (hunk 6 above)
Comment (abeddow91) [reply article/app/model/LiveBlogHelpers.scala:105]: addressed in ab/pinned-block-rename

File: article/app/views/liveblog/liveBlogBodyContent.scala.html (hunk 12 above)
Comment (abeddow91) [reply article/app/views/liveblog/liveBlogBodyContent.scala.html:39]: addresed in ab/pinned-block-rename

File: article/app/views/liveblog/pinnedPost.scala.html (hunk 13 above)
Comment (abeddow91) [reply article/app/views/liveblog/pinnedPost.scala.html:1]: addresed in ab/pinned-block-rename

File: article/app/views/liveblog/liveBlogBody.scala.html (hunk 11 above)
Comment (abeddow91) [reply article/app/views/liveblog/liveBlogBody.scala.html:109]: removed

**Review by alinaboghiu — APPROVED** [review]

**Pull request merged**